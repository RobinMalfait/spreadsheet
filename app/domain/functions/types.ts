import { type EvaluationResult, EvaluationResultKind } from '~/domain/evaluation'

export function AS_NUMBER(
  value?: EvaluationResult,
  extra?: EvaluationResult,
): EvaluationResult {
  if (value === undefined) {
    return {
      kind: EvaluationResultKind.ERROR,
      value: 'AS_NUMBER() expects a value as the first argument, got <nothing>',
    }
  }

  if (extra) {
    return {
      kind: EvaluationResultKind.ERROR,
      value: `AS_NUMBER() does not take more than one argument, got ${extra.value}`,
    }
  }

  switch (value.kind) {
    case EvaluationResultKind.ERROR: // No need to convert an error to a number
    case EvaluationResultKind.NUMBER: // No need to convert a number to a number
      return value

    case EvaluationResultKind.BOOLEAN:
      return { kind: EvaluationResultKind.NUMBER, value: value.value ? 1 : 0 }

    case EvaluationResultKind.DATETIME:
      return { kind: EvaluationResultKind.NUMBER, value: value.value.getTime() }

    case EvaluationResultKind.STRING: {
      let asNumber = Number(value.value)
      if (Number.isNaN(asNumber)) {
        return {
          kind: EvaluationResultKind.ERROR,
          value: `AS_NUMBER() expects a number, got ${value.value}`,
        }
      }

      return { kind: EvaluationResultKind.NUMBER, value: asNumber }
    }

    default:
      value satisfies never
      return value
  }
}
