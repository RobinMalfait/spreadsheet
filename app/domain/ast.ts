import type { Location } from '~/domain/expression'

export enum AstKind {
  CELL = 'CELL',
  RANGE = 'RANGE',
  FUNCTION = 'FUNCTION',
  NUMBER_LITERAL = 'NUMBER_LITERAL',
  STRING_LITERAL = 'STRING_LITERAL',
  BINARY_EXPRESSION = 'BINARY_EXPRESSION',
}

interface Span {
  span: {
    start: number
    end: number
  }
}

export interface AstCell extends Span {
  kind: AstKind.CELL
  name: string
  raw: string
  loc: Location
}

export interface AstCellRange extends Span {
  kind: AstKind.RANGE
  raw: string
  start: AstCell
  end: AstCell
}

export interface AstFunction extends Span {
  kind: AstKind.FUNCTION
  name: string
  args: AST[]
}

interface AstNumberLiteral extends Span {
  kind: AstKind.NUMBER_LITERAL
  value: number
}

interface AstStringLiteral extends Span {
  kind: AstKind.STRING_LITERAL
  value: string
}

interface AstBinaryExpression extends Span {
  kind: AstKind.BINARY_EXPRESSION
  operator: BinaryExpressionOperator
  lhs: AST
  rhs: AST
}

export type AST =
  | AstCell
  | AstCellRange
  | AstFunction
  | AstNumberLiteral
  | AstStringLiteral
  | AstBinaryExpression

export enum BinaryExpressionOperator {
  // Math operators
  EXPONENT = 'EXPONENT',
  MULTIPLY = 'MULTIPLY',
  ADD = 'ADD',
  SUBTRACT = 'SUBTRACT',
  DIVIDE = 'DIVIDE',

  // Comparison operators
  EQUALS = 'EQUALS',
  NOT_EQUALS = 'NOT_EQUALS',
  LESS_THAN = 'LESS_THAN',
  LESS_THAN_EQUALS = 'LESS_THAN_EQUALS',
  GREATER_THAN = 'GREATER_THAN',
  GREATER_THAN_EQUALS = 'GREATER_THAN_EQUALS',
}
