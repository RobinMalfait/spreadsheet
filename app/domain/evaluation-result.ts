export enum EvaluationResultKind {
  ERROR = 'ERROR',
  EMPTY = 'EMPTY',
  NUMBER = 'NUMBER',
  STRING = 'STRING',
  BOOLEAN = 'BOOLEAN',
  DATETIME = 'DATETIME',
}

export type EvaluationResult =
  | { kind: EvaluationResultKind.ERROR; value: string }
  | { kind: EvaluationResultKind.EMPTY; value: string }
  | { kind: EvaluationResultKind.NUMBER; value: number }
  | { kind: EvaluationResultKind.STRING; value: string }
  | { kind: EvaluationResultKind.BOOLEAN; value: boolean }
  | { kind: EvaluationResultKind.DATETIME; value: Date; date: boolean; time: boolean }
