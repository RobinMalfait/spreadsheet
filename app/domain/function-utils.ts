import type { EvaluationResult } from '~/domain/evaluation-result'
import { parse } from '~/domain/signature/parser'
import { tokenize } from '~/domain/signature/tokenizer'
import { validate } from '~/domain/signature/validate'

export function expose<
  T extends EvaluationResult,
  R extends EvaluationResult | EvaluationResult[],
>(signature: string, handle: (...args: T[]) => R): (...args: T[]) => R {
  let sig = parse(tokenize(signature))

  return Object.assign(
    (...args: T[]) => {
      let errors = validate(sig, args)
      if (errors) return errors as R

      // Handle
      return handle(...args) as R
    },
    { signature: sig },
  )
}
