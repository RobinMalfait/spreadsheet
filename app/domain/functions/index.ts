export * from './binary'
export * from './date'
export * from './logic'
export * from './lookup'
export * from './math'
export * from './privileged'
export * from './sequence'
export * from './statistics'
export * from './text'
export * from './types'
