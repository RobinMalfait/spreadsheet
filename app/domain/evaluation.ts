import { format } from 'date-fns'
import { type AST, AstKind, BinaryExpressionOperator } from '~/domain/ast'
import * as functions from '~/domain/functions'
import type { Spreadsheet } from '~/domain/spreadsheet'
import { expandRange } from '~/domain/walk-ast'

export enum EvaluationResultKind {
  NUMBER = 'NUMBER',
  STRING = 'STRING',
  BOOLEAN = 'BOOLEAN',
  DATETIME = 'DATETIME',
}

export type EvaluationResult =
  | { kind: EvaluationResultKind.NUMBER; value: number }
  | { kind: EvaluationResultKind.STRING; value: string }
  | { kind: EvaluationResultKind.BOOLEAN; value: boolean }
  | { kind: EvaluationResultKind.DATETIME; value: Date; date: boolean; time: boolean }

export function evaluateExpression(
  ast: AST,
  spreadsheet: Spreadsheet,
): EvaluationResult[] {
  switch (ast.kind) {
    case AstKind.NUMBER_LITERAL:
      return [{ kind: EvaluationResultKind.NUMBER, value: ast.value }]

    case AstKind.STRING_LITERAL: {
      if (ast.value.trim() === '') {
        return [{ kind: EvaluationResultKind.STRING, value: ast.value }]
      }

      // Try to coerce the string to a number, and if it works, return a number
      // instead of a string
      let asNumber = Number(ast.value)
      if (!Number.isNaN(asNumber)) {
        return [{ kind: EvaluationResultKind.NUMBER, value: asNumber }]
      }

      return [{ kind: EvaluationResultKind.STRING, value: ast.value }]
    }

    case AstKind.CELL:
      return spreadsheet.evaluate(ast.name)

    case AstKind.RANGE: {
      let out = []
      for (let cell of expandRange(ast)) {
        for (let child of spreadsheet.evaluate(cell)) {
          out.push(child)
        }
      }
      return out
    }

    case AstKind.BINARY_EXPRESSION: {
      let lhs = evaluateExpression(ast.lhs, spreadsheet)
      let rhs = evaluateExpression(ast.rhs, spreadsheet)

      if (lhs.length !== 1 || rhs.length !== 1) {
        throw new Error('Expected a single result from each side')
      }

      let left = lhs[0]
      let right = rhs[0]

      if (
        left?.kind === EvaluationResultKind.NUMBER &&
        right?.kind === EvaluationResultKind.NUMBER
      ) {
        switch (ast.operator) {
          // Math operators
          case BinaryExpressionOperator.EXPONENT:
            return [functions.POWER(left, right)]
          case BinaryExpressionOperator.MULTIPLY:
            return [functions.MULTIPLY(left, right)]
          case BinaryExpressionOperator.DIVIDE:
            return [functions.DIVIDE(left, right)]
          case BinaryExpressionOperator.ADD:
            return [functions.SUM(left, right)]
          case BinaryExpressionOperator.SUBTRACT:
            return [functions.SUBTRACT(left, right)]

          // Comparison operators
          case BinaryExpressionOperator.EQUALS:
            return left.value === right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.NOT_EQUALS:
            return left.value !== right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.LESS_THAN:
            return left.value < right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.LESS_THAN_EQUALS:
            return left.value <= right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.GREATER_THAN:
            return left.value > right.value ? [functions.TRUE()] : [functions.FALSE()]
          case BinaryExpressionOperator.GREATER_THAN_EQUALS:
            return left.value >= right.value ? [functions.TRUE()] : [functions.FALSE()]
          default:
            ast.operator satisfies never
        }
      }

      throw new Error(`Invalid operation \`${ast.kind}\``)
    }

    case AstKind.FUNCTION: {
      if (!Object.hasOwn(functions, ast.name)) {
        throw new Error(`Unknown function \`${ast.name}\``)
      }

      let fn = functions[ast.name as keyof typeof functions]
      let args = ast.args.flatMap((arg) => evaluateExpression(arg, spreadsheet))
      // @ts-expect-error Some functions have a different arity, but we're not
      // checking that here.
      let result = fn(...args)
      if (Array.isArray(result)) {
        return result
      }
      return [result]
    }
  }
}

export function printEvaluationResult(result: EvaluationResult): string {
  switch (result.kind) {
    case EvaluationResultKind.NUMBER:
      return result.value.toString()
    case EvaluationResultKind.STRING:
      return result.value
    case EvaluationResultKind.BOOLEAN:
      return result.value ? 'TRUE' : 'FALSE'
    case EvaluationResultKind.DATETIME:
      if (result.date && result.time) return format(result.value, 'yyyy-MM-dd HH:mm:ss')
      if (result.date) return format(result.value, 'yyyy-MM-dd')
      return format(result.value, 'HH:mm:ss')
    default:
      result satisfies never
      return ''
  }
}
