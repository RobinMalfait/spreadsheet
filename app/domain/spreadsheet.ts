import { type AST, AstKind } from '~/domain/ast'
import { type EvaluationResult, evaluateExpression } from '~/domain/evaluation'
import { parse } from '~/domain/expression'
import * as functions from '~/domain/functions'
import { tokenize } from '~/domain/tokenizer'
import { WalkAction, walk } from '~/domain/walk-ast'
import { DefaultMap } from '~/utils/default-map'

export enum ComputationResultKind {
  VALUE = 'VALUE',
  ERROR = 'ERROR',
}

type ComputationValue = {
  kind: ComputationResultKind.VALUE
  value: EvaluationResult
}

type ComputationError = {
  kind: ComputationResultKind.ERROR
  short: string
  message: string
}

export class Spreadsheet {
  // Track each individual cell and it's contents. AST is pre-parsed.
  private cells = new Map<string, [raw: string, ast: AST]>()

  // Track all dependencies for each cell.
  private _dependencies = new DefaultMap<string, Set<string>>(() => new Set<string>())

  private evaluationCache = new Map<string, EvaluationResult[]>()

  get cellNames() {
    return Array.from(this.cells.keys())
  }

  has(cell: string): boolean {
    return this.cells.has(cell)
  }

  get(cell: string): string {
    let result = this.cells.get(cell)
    if (result) {
      return result[0]
    }
    return ''
  }

  set(cell: string, value: string) {
    // Reset state
    this.cells.delete(cell)

    // Clear existing dependencies for this cell
    let dependencies = this._dependencies.get(cell)
    dependencies.clear()

    // Clear the full evaluation cache. Very naive, but let's re-compute
    // everything the moment _something_ changes.
    this.evaluationCache.clear()

    // Value must be filled in
    if (value.trim() === '') return

    // Expression should exist
    let expression = value[0] === '=' ? value.slice(1) : `"${value}"`
    if (expression.trim() === '') return

    let tokens = tokenize(expression)
    let ast = parse(tokens)

    // Track all references in the AST
    walk([ast], (node) => {
      if (node.kind === AstKind.CELL) {
        dependencies.add(node.name)
      }

      return WalkAction.Continue
    })

    this.cells.set(cell, [value, ast])
  }

  functions() {
    let collator = new Intl.Collator('en', { sensitivity: 'base' })
    return Object.keys(functions).sort(collator.compare)
  }

  dependencies(cell: string): Set<string> {
    return new Set(this._dependencies.get(cell))
  }

  compute(cell: string): ComputationValue | ComputationError | null {
    try {
      let evaluationResult = this.evaluate(cell)
      if (evaluationResult.length > 1) {
        throw Object.assign(
          new Error(`Expected a single result, got ${evaluationResult.length}`),
          { short: '#VALUE!' },
        )
      }

      if (evaluationResult.length === 1) {
        return {
          kind: ComputationResultKind.VALUE,
          // biome-ignore lint/style/noNonNullAssertion: we already verified that the evaluationResult has at least one element.
          value: evaluationResult[0]!,
        }
      }

      return null
    } catch (err: unknown) {
      return {
        kind: ComputationResultKind.ERROR,
        // @ts-expect-error This is fineâ€¦
        short: err?.short ?? 'Error',
        message: (err as Error).message,
      }
    }
  }

  evaluate(cell: string): EvaluationResult[] {
    let result = this.cells.get(cell)
    if (!result) return []

    let cached = this.evaluationCache.get(cell)
    if (cached) return cached

    // TODO: Should this be moved to the `set` method?
    if (result[1].kind === AstKind.RANGE) {
      throw Object.assign(new Error('Cannot reference a range to a cell'), {
        short: '#VALUE!',
      })
    }

    // Verify references
    let dependencies = this._dependencies.get(cell)
    if (dependencies.size > 0) {
      let handled = new Set<string>()
      let todo = Array.from(dependencies)

      // Track where we came from
      let previous = cell

      while (todo.length > 0) {
        // biome-ignore lint/style/noNonNullAssertion: We verified that there is at least a single element in the while condition above.
        let next = todo.shift()!

        // No need to re-verify cells we've already handled
        if (handled.has(next)) continue
        handled.add(next)

        if (next === cell) {
          throw Object.assign(
            new Error(`Circular reference detected in cell ${previous}`),
            { short: '#REF!' },
          )
        }

        if (this.cells.has(next)) {
          // Track where we came from
          previous = next

          // Check transitive dependencies
          for (let other of this._dependencies.get(next)) {
            if (!handled.has(other)) {
              todo.push(other)
            }
          }
        }
      }
    }

    let out = evaluateExpression(result[1], this)
    this.evaluationCache.set(cell, out)
    return out
  }
}
