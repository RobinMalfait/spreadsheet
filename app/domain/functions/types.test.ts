import { beforeAll, describe, expect, it, vi } from 'vitest'
import { Spreadsheet } from '~/domain/spreadsheet'
import { visualizeSpreadsheet } from '~/test/utils'
import { exampleTests } from '~/test/utils'
import * as functions from './types'
await exampleTests(functions)

beforeAll(() => {
  vi.setSystemTime(new Date(2013, 0, 21, 8, 15, 20))
})

describe('AS_NUMBER()', () => {
  it('should error when the argument is not provided', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=AS_NUMBER()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: AS_NUMBER(value: T) Argument \`value\` was not provided
      "
    `)
  })

  it('should error when providing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=AS_NUMBER(123, 456)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: AS_NUMBER(value: T) takes at most 1 argument, got 2
      "
    `)
  })

  it('should convert any type to a number if possible', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', 'Input:')
    spreadsheet.set('A2', 'Output:')
    spreadsheet.set('A3', 'Type:')

    spreadsheet.set('B1', '=123')
    spreadsheet.set('B2', '=AS_NUMBER(B1)')
    spreadsheet.set('B3', '=TYPE(B2)')

    spreadsheet.set('C1', '=TRUE()')
    spreadsheet.set('C2', '=AS_NUMBER(C1)')
    spreadsheet.set('C3', '=TYPE(C2)')

    spreadsheet.set('D1', '=FALSE()')
    spreadsheet.set('D2', '=AS_NUMBER(D1)')
    spreadsheet.set('D3', '=TYPE(D2)')

    spreadsheet.set('E1', '="123"')
    spreadsheet.set('E2', '=AS_NUMBER(E1)')
    spreadsheet.set('E3', '=TYPE(E2)')

    spreadsheet.set('F1', '="invalid text"')
    spreadsheet.set('F2', '=AS_NUMBER(F1)')
    spreadsheet.set('F3', '=TYPE(F2)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────┬──────────┬──────────┬──────────┬──────────┬────────────────┐
      │   │ A       │ B        │ C        │ D        │ E        │ F              │
      ├───┼─────────┼──────────┼──────────┼──────────┼──────────┼────────────────┤
      │ 1 │ Input:  │ 123      │ TRUE     │ FALSE    │ "123"    │ "invalid text" │
      ├───┼─────────┼──────────┼──────────┼──────────┼──────────┼────────────────┤
      │ 2 │ Output: │ 123      │ 1        │ 0        │ 123      │ Error          │
      ├───┼─────────┼──────────┼──────────┼──────────┼──────────┼────────────────┤
      │ 3 │ Type:   │ "number" │ "number" │ "number" │ "number" │ "error"        │
      └───┴─────────┴──────────┴──────────┴──────────┴──────────┴────────────────┘

      Errors:

      · F2: AS_NUMBER() expects a number, got invalid text
      "
    `)
  })
})

describe('AS_STRING()', () => {
  it('should error when the argument is not provided', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=AS_STRING()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: AS_STRING(value: T) Argument \`value\` was not provided
      "
    `)
  })

  it('should error when providing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=AS_STRING(123, 456)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: AS_STRING(value: T) takes at most 1 argument, got 2
      "
    `)
  })

  it('should convert any type to a string if possible', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', 'Input:')
    spreadsheet.set('A2', 'Output:')
    spreadsheet.set('A3', 'Type:')

    spreadsheet.set('B1', '=123')
    spreadsheet.set('B2', '=AS_STRING(B1)')
    spreadsheet.set('B3', '=TYPE(B2)')

    spreadsheet.set('C1', '=TRUE()')
    spreadsheet.set('C2', '=AS_STRING(C1)')
    spreadsheet.set('C3', '=TYPE(C2)')

    spreadsheet.set('D1', '=FALSE()')
    spreadsheet.set('D2', '=AS_STRING(D1)')
    spreadsheet.set('D3', '=TYPE(D2)')

    spreadsheet.set('E1', '="hello world"')
    spreadsheet.set('E2', '=AS_STRING(E1)')
    spreadsheet.set('E3', '=TYPE(E2)')

    spreadsheet.set('F1', '=NOW()')
    spreadsheet.set('F2', '=AS_STRING(F1)')
    spreadsheet.set('F3', '=TYPE(F2)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────┬──────────┬──────────┬──────────┬───────────────┬───────────────────────┐
      │   │ A       │ B        │ C        │ D        │ E             │ F                     │
      ├───┼─────────┼──────────┼──────────┼──────────┼───────────────┼───────────────────────┤
      │ 1 │ Input:  │ 123      │ TRUE     │ FALSE    │ "hello world" │ 2013-01-21 08:15:20   │
      ├───┼─────────┼──────────┼──────────┼──────────┼───────────────┼───────────────────────┤
      │ 2 │ Output: │ "123"    │ "TRUE"   │ "FALSE"  │ "hello world" │ "2013-01-21 08:15:20" │
      ├───┼─────────┼──────────┼──────────┼──────────┼───────────────┼───────────────────────┤
      │ 3 │ Type:   │ "string" │ "string" │ "string" │ "string"      │ "string"              │
      └───┴─────────┴──────────┴──────────┴──────────┴───────────────┴───────────────────────┘
      "
    `)
  })
})

describe('AS_BOOLEAN()', () => {
  it('should error when the argument is not provided', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=AS_BOOLEAN()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: AS_BOOLEAN(value: T) Argument \`value\` was not provided
      "
    `)
  })

  it('should error when providing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=AS_BOOLEAN(123, 456)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: AS_BOOLEAN(value: T) takes at most 1 argument, got 2
      "
    `)
  })

  it('should convert any type to a string if possible', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', 'Input:')
    spreadsheet.set('A2', 'Output:')
    spreadsheet.set('A3', 'Type:')

    spreadsheet.set('B1', '=123')
    spreadsheet.set('B2', '=AS_BOOLEAN(B1)')
    spreadsheet.set('B3', '=TYPE(B2)')

    spreadsheet.set('C1', '=0')
    spreadsheet.set('C2', '=AS_BOOLEAN(C1)')
    spreadsheet.set('C3', '=TYPE(C2)')

    spreadsheet.set('D1', '=TRUE()')
    spreadsheet.set('D2', '=AS_BOOLEAN(D1)')
    spreadsheet.set('D3', '=TYPE(D2)')

    spreadsheet.set('E1', '=FALSE()')
    spreadsheet.set('E2', '=AS_BOOLEAN(E1)')
    spreadsheet.set('E3', '=TYPE(E2)')

    spreadsheet.set('F1', '=""')
    spreadsheet.set('F2', '=AS_BOOLEAN(F1)')
    spreadsheet.set('F3', '=TYPE(F2)')

    spreadsheet.set('G1', '="hello world"')
    spreadsheet.set('G2', '=AS_BOOLEAN(G1)')
    spreadsheet.set('G3', '=TYPE(G2)')

    spreadsheet.set('H1', '=NOW()')
    spreadsheet.set('H2', '=AS_BOOLEAN(H1)')
    spreadsheet.set('H3', '=TYPE(H2)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────┬───────────┬───────────┬───────────┬───────────┬───────────┬───────────────┬─────────────────────┐
      │   │ A       │ B         │ C         │ D         │ E         │ F         │ G             │ H                   │
      ├───┼─────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────────┼─────────────────────┤
      │ 1 │ Input:  │ 123       │ 0         │ TRUE      │ FALSE     │ ""        │ "hello world" │ 2013-01-21 08:15:20 │
      ├───┼─────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────────┼─────────────────────┤
      │ 2 │ Output: │ TRUE      │ FALSE     │ TRUE      │ FALSE     │ FALSE     │ TRUE          │ TRUE                │
      ├───┼─────────┼───────────┼───────────┼───────────┼───────────┼───────────┼───────────────┼─────────────────────┤
      │ 3 │ Type:   │ "boolean" │ "boolean" │ "boolean" │ "boolean" │ "boolean" │ "boolean"     │ "boolean"           │
      └───┴─────────┴───────────┴───────────┴───────────┴───────────┴───────────┴───────────────┴─────────────────────┘
      "
    `)
  })
})
