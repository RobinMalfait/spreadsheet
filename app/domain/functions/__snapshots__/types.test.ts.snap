// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@example tests > AS_BOOLEAN > AS_BOOLEAN("0") 1`] = `
"
┌───┬───────┐
│   │ A     │
├───┼───────┤
│ 1 │ FALSE │
└───┴───────┘
"
`;

exports[`@example tests > AS_BOOLEAN > AS_BOOLEAN("123") 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ TRUE │
└───┴──────┘
"
`;

exports[`@example tests > AS_BOOLEAN > AS_BOOLEAN(0) 1`] = `
"
┌───┬───────┐
│   │ A     │
├───┼───────┤
│ 1 │ FALSE │
└───┴───────┘
"
`;

exports[`@example tests > AS_BOOLEAN > AS_BOOLEAN(1) 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ TRUE │
└───┴──────┘
"
`;

exports[`@example tests > AS_BOOLEAN > AS_BOOLEAN(FALSE()) 1`] = `
"
┌───┬───────┐
│   │ A     │
├───┼───────┤
│ 1 │ FALSE │
└───┴───────┘
"
`;

exports[`@example tests > AS_BOOLEAN > AS_BOOLEAN(NOW()) 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ TRUE │
└───┴──────┘
"
`;

exports[`@example tests > AS_BOOLEAN > AS_BOOLEAN(TRUE()) 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ TRUE │
└───┴──────┘
"
`;

exports[`@example tests > AS_BOOLEANS(0, 1, "123", "0", TRUE(), FALSE(), NOW()) 1`] = `
"
┌───┬───────┬──────┬──────┬───────┬──────┬───────┬──────┐
│   │ A     │ B    │ C    │ D     │ E    │ F     │ G    │
├───┼───────┼──────┼──────┼───────┼──────┼───────┼──────┤
│ 1 │ FALSE │ TRUE │ TRUE │ FALSE │ TRUE │ FALSE │ TRUE │
└───┴───────┴──────┴──────┴───────┴──────┴───────┴──────┘
"
`;

exports[`@example tests > AS_CHAR > AS_CHAR(65) 1`] = `
"
┌───┬─────┐
│   │ A   │
├───┼─────┤
│ 1 │ "A" │
└───┴─────┘
"
`;

exports[`@example tests > AS_CHAR > AS_CHAR(97) 1`] = `
"
┌───┬─────┐
│   │ A   │
├───┼─────┤
│ 1 │ "a" │
└───┴─────┘
"
`;

exports[`@example tests > AS_NUMBER > AS_NUMBER("123") 1`] = `
"
┌───┬─────┐
│   │ A   │
├───┼─────┤
│ 1 │ 123 │
└───┴─────┘
"
`;

exports[`@example tests > AS_NUMBER > AS_NUMBER("million") 1`] = `
"
┌───┬───────┐
│   │ A     │
├───┼───────┤
│ 1 │ Error │
└───┴───────┘

Errors:

· A1: AS_NUMBER() expects a number, got million
"
`;

exports[`@example tests > AS_NUMBER > AS_NUMBER(1) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 1 │
└───┴───┘
"
`;

exports[`@example tests > AS_NUMBER > AS_NUMBER(FALSE()) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 0 │
└───┴───┘
"
`;

exports[`@example tests > AS_NUMBER > AS_NUMBER(NOW()) 1`] = `
"
┌───┬───────────────┐
│   │ A             │
├───┼───────────────┤
│ 1 │ 1358752520000 │
└───┴───────────────┘
"
`;

exports[`@example tests > AS_NUMBER > AS_NUMBER(TRUE()) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 1 │
└───┴───┘
"
`;

exports[`@example tests > AS_NUMBERS(1, "123", "million", TRUE(), FALSE(), NOW()) 1`] = `
"
┌───┬───┬─────┬───────┬───┬───┬───────────────┐
│   │ A │ B   │ C     │ D │ E │ F             │
├───┼───┼─────┼───────┼───┼───┼───────────────┤
│ 1 │ 1 │ 123 │ Error │ 1 │ 0 │ 1358752520000 │
└───┴───┴─────┴───────┴───┴───┴───────────────┘

Errors:

· C1: AS_NUMBER() expects a number, got million
"
`;

exports[`@example tests > AS_STRING > AS_STRING("123") 1`] = `
"
┌───┬───────┐
│   │ A     │
├───┼───────┤
│ 1 │ "123" │
└───┴───────┘
"
`;

exports[`@example tests > AS_STRING > AS_STRING(1) 1`] = `
"
┌───┬─────┐
│   │ A   │
├───┼─────┤
│ 1 │ "1" │
└───┴─────┘
"
`;

exports[`@example tests > AS_STRING > AS_STRING(FALSE()) 1`] = `
"
┌───┬─────────┐
│   │ A       │
├───┼─────────┤
│ 1 │ "FALSE" │
└───┴─────────┘
"
`;

exports[`@example tests > AS_STRING > AS_STRING(NOW()) 1`] = `
"
┌───┬───────────────────────┐
│   │ A                     │
├───┼───────────────────────┤
│ 1 │ "2013-01-21 08:15:20" │
└───┴───────────────────────┘
"
`;

exports[`@example tests > AS_STRING > AS_STRING(TIME()) 1`] = `
"
┌───┬────────────┐
│   │ A          │
├───┼────────────┤
│ 1 │ "08:15:20" │
└───┴────────────┘
"
`;

exports[`@example tests > AS_STRING > AS_STRING(TRUE()) 1`] = `
"
┌───┬────────┐
│   │ A      │
├───┼────────┤
│ 1 │ "TRUE" │
└───┴────────┘
"
`;

exports[`@example tests > AS_STRINGS(1, "123", TRUE(), FALSE(), NOW(), TIME()) 1`] = `
"
┌───┬─────┬───────┬────────┬─────────┬───────────────────────┬────────────┐
│   │ A   │ B     │ C      │ D       │ E                     │ F          │
├───┼─────┼───────┼────────┼─────────┼───────────────────────┼────────────┤
│ 1 │ "1" │ "123" │ "TRUE" │ "FALSE" │ "2013-01-21 08:15:20" │ "08:15:20" │
└───┴─────┴───────┴────────┴─────────┴───────────────────────┴────────────┘
"
`;

exports[`@example tests > TYPE > TYPE("hello") 1`] = `
"
┌───┬──────────┐
│   │ A        │
├───┼──────────┤
│ 1 │ "string" │
└───┴──────────┘
"
`;

exports[`@example tests > TYPE > TYPE(1) 1`] = `
"
┌───┬──────────┐
│   │ A        │
├───┼──────────┤
│ 1 │ "number" │
└───┴──────────┘
"
`;

exports[`@example tests > TYPE > TYPE(B1) 1`] = `
"
┌───┬─────────┐
│   │ A       │
├───┼─────────┤
│ 1 │ "empty" │
└───┴─────────┘
"
`;

exports[`@example tests > TYPE > TYPE(NOW()) 1`] = `
"
┌───┬────────────┐
│   │ A          │
├───┼────────────┤
│ 1 │ "datetime" │
└───┴────────────┘
"
`;

exports[`@example tests > TYPE > TYPE(TRUE()) 1`] = `
"
┌───┬───────────┐
│   │ A         │
├───┼───────────┤
│ 1 │ "boolean" │
└───┴───────────┘
"
`;

exports[`@example tests > TYPE > TYPE(UNKNOWN_FUNCTION()) 1`] = `
"
┌───┬─────────┐
│   │ A       │
├───┼─────────┤
│ 1 │ "error" │
└───┴─────────┘
"
`;
